### PR TITLE
feat(toast): add 'swipeGesture' property to dismiss toasts via swipe(…

### PR DESCRIPTION
**Pull request by evgeniy-skakun**
…TakumaKira) - expanded

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

https://github.com/ionic-team/ionic-framework/issues/21769

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by liamdebeasi** [comment]
Thanks for the PR. What is the difference between this PR and the previous toast swipe gesture PR? (https://github.com/ionic-team/ionic-framework/pull/22133).

**Comment by evgeniy-skakun** [comment]
@liamdebeasi  Please check the extended version of the PR [ @TakumaKira ] https://github.com/ionic-team/ionic-framework/pull/22133

**Comment by evgeniy-skakun** [comment]
@liamdebeasi  Added "vertical" and "horizontal" swipe

**Comment by liamdebeasi** [comment]
What is the difference between a horizontal swipe and a swipe from the left or right?

**Comment by evgeniy-skakun** [comment]
@liamdebeasi You can swipe in both directions instead of just one.

**Comment by liamdebeasi** [comment]
Do you have any use cases for this? I wonder if it makes sense to just have the options be `'horizontal' | 'vertical'` and determine the direction based upon where the toast was presented from.

For example, if the toast was presented from the bottom and I had `swipeGesture="vertical"` set, I can't imagine there is a use case where I would want to swipe up to dismiss. I would expect that I would just swipe down to dismiss since it is closest to the bottom edge of the screen. 

Similarly, is there any benefit to only letting someone swipe to the left to dismiss versus being able to swipe to the left or the right?

**Comment by evgeniy-skakun** [comment]
`vertical swipe` will be useful if the toast is placed in the center of the screen.

`horizontal swipe` can be useful if the application is used by left-handers and right-handers.

@liamdebeasi In any case, if you do not see any use in them, then they can be removed.

**Comment by liamdebeasi** [comment]
Thanks for the clarification, that makes sense. I think in that case we can infer the direction based upon the `position` prop value.

If `position="bottom"` then swiping down would dismiss the toast.
If `position="top"` then swiping up would dismiss the toast.
If `position="middle` then swiping up or down would dismiss the toast.

Do you have any examples of the horizontal swipe in existing applications? I like the vertical swipe, but my concern with the horizontal swipe is that it is does not seem to exist in the Material Design spec.

**Comment by evgeniy-skakun** [comment]
https://user-images.githubusercontent.com/58041146/113181917-f92b2b80-925a-11eb-9e01-b83b3fe38b45.mp4

Samsung allows you to swipe in both directions

**Comment by liamdebeasi** [comment]
Those looks like Android-specific notification center components, not a toast/snackbar. Do you have any examples of this horizontal gesture with a toast/snackbar? 

For example, the Snackbar Flutter demo on https://material.io/components/snackbars only lets you do a vertical swipe.

**Comment by evgeniy-skakun** [comment]
@liamdebeasi  I don't have an example app.

I can remove the `horizontal swipe` if you think it doesn't make sense.

**Comment by liamdebeasi** [comment]
No problem, was just curious you knew of an example. I do see the benefit of this feature, so I think this is something we should add.

This is a new API, so it would need to go into a minor release. We are focusing on patch releases and the new major version of Ionic Framework at the moment, but I can add this to our project board and we will do a full review as we get closer to planning a new minor release.

File: core/src/components/toast/toast.tsx [hunk 1]
```diff
@@ -224,6 +246,147 @@ export class Toast implements ComponentInterface, OverlayInterface {
...
+            () => shouldComplete && this.dismiss()
+          )
+          .progressEnd((shouldComplete) ? 1 : 0, stepValue, 300);
+      }
+    });
+    this.gesture.enable(true);
+  }
+
+  private clamp = (val: number) => {
+    if (val > 1) {
```
Comment (lodenrogue) [inline core/src/components/toast/toast.tsx:382]: Can this be simplified?

```
return val > 1 ? 1 : Math.max(0, val);
```

**Review by lodenrogue — CHANGES_REQUESTED** [review]

**Comment by reslear** [comment]
any news?

**Comment by katche70** [comment]
Hi Ionic Team, 
It would be great if this feature comes in the official Ionic release

**Comment by thekhegay** [comment]
Any updates?

**Comment by bstrasburg** [comment]
Any updates ????

**Comment by philmmoore** [comment]
I'd find this useful in my apps too :)

**Comment by EinfachHans** [comment]
can't wait to se this in the framwork

**Comment by katche70** [comment]
Hey whats going on here with this open PR ? Nobody form ionic Team here to add this in official ionic release. Would by nice to have this feature, because it would be feel more native for ios user. Push Messages from ios simply can swipe up

**Comment by liamdebeasi** [comment]
Hi everyone,

The team met a few weeks ago to discuss this PR, and we are interested in adding this feature. The next step is for the team to do a feature design document. This is where we determine the scope, requirements, and architecture of the feature to help ship a reliable API. Once that has been approved, someone from the team will review the PR and help make any changes necessary to get this PR ready to merge.

Our primary focus right now is getting Ionic 7.0 into beta, which is why you have not seen much movement on this PR over the past few weeks. While I do not have a timeline for when this will merge, we recognize that this is an important feature for Ionic developers. Let me know if there are any questions. Thanks!

**Comment by benmarsh** [comment]
This is great news - thanks for the update @liamdebeasi :)

**Comment by rsammelson** [comment]
Is there any update on this being completed? Would a new PR with conflicts resolved be useful to that process?

**Comment by sarevok89** [comment]
Is there any update on this? The PR has been opened more than 2 years above 😞

**Comment by Sigmale1000** [comment]
Any updates on the feature, since Ionic 7 is out?

**Comment by liamdebeasi** [comment]
Hey there! Thanks for contributing this PR, and apologies for the delay in replying here. This feature was added to Ionic in https://github.com/ionic-team/ionic-framework/pull/28442. The work you did here helped us build out the final implementation, so we've given you co-author credit in the commit that was merged.

This feature will be available in an upcoming minor release of Ionic. I am going to close this, and thanks again for contributing to Ionic!

**Pull request closed without merging**